### PR TITLE
Update proteus for release v1.0

### PR DESCRIPTION
**Pull request by gshiroma**
This PR updates proteus for release v1.0. Changes:
- Rename metadata fields imported from HLS product from `SPATIAL_COVERAGE` and `CLOUD_COVERAGE` to `INPUT_HLS_PRODUCT_SPATIAL_COVERAGE` and `INPUT_HLS_PRODUCT_CLOUD_COVERAGE`, respectively;
- Recompute and store the new metadata fields `SPATIAL_COVERAGE` and `CLOUD_COVERAGE` considering fill value and ocean masked data as invalid data;
- Save new metadata fields describing processing parameters used for generating the DSWx-HLS product: `SHADOW_MASKING_ALGORITHM`, `MIN_SLOPE_ANGLE`, `MAX_SUN_LOCAL_INC_ANGLE`, `MASK_ADJACENT_TO_CLOUD_MODE`, `FOREST_MASK_LANDCOVER_CLASSES`, and `OCEAN_MASKING_SHORELINE_DISTANCE_KM`;
- Update version tag from `v.0.5.2` to `v1.0`.
- Add high-aerosol masking.

### CODE REVIEW

File: src/proteus/dswx_hls.py
Comment (nemo794) [inline src/proteus/dswx_hls.py:4274]: I've gone back and forth about this preliminary cloud layer. It was really confusing to me until I spent a bit of time with the code, and I really do not like splitting the logic for the CLOUD layer into two separate places; the docstrings are confusing/redundant, and creates extra steps that need to be tracked and followed up on later in the code.

What about a new approach: 

1) Put all logic for creating the CLOUD layer (cloud+shadow+adj+snow) into a new function, e.g. `_compute_cloud_layer()`.
2) Call that function at this point in the code.
3) When computing the statistic of `CLOUD_COVERAGE`, it would become something like:
```
n_cloud_and_valid = np.sum(((preliminary_cloud_layer) != 0) & 
                                                    (preliminary_cloud_layer) != SNOW) & 
                                                    (valid_array))
```
4) Later, turn the function `_compute_and_apply_cloud_layer()` into simply `_compute_wtr_2_layer(wtr_1_layer=..., cloud_layer=...)`

This approach would greatly simplify, streamline, and modularize the code, reduce docstring redundancies and errors, and help future maintainers see exactly where/how each layer is being computed.

File: src/proteus/dswx_hls.py
Comment (nemo794) [inline src/proteus/dswx_hls.py:4450]: Please see previous comment about how to better modularize the code.

File: src/proteus/dswx_hls.py
Comment (nemo794) [inline src/proteus/dswx_hls.py:3522]: For the metadata, should class 201 (evergreen) also be included in this list? Even though this class is hardcoded into the algorithm, it is one of the Copernius forest classes which is being masked.

File: src/proteus/dswx_hls.py
Comment (nemo794) [inline src/proteus/dswx_hls.py:3525]: Should this only be a number if a shoreline shape file is provided?

If there's no shoreline file, then we do not create an ocean mask, correct? In which case, should this be marked with an e.g. `-`, similar to how `MIN_SLOPE_ANGLE` is handled depending on `SHADOW_MASKING_ALGORITHM` used?

File: src/proteus/dswx_hls.py
Comment (nemo794) [inline src/proteus/dswx_hls.py]: I left this comment elsewhere, but shouldn't class 201 (evergreen) also be included in this list for the metadata?

File: src/proteus/schemas/dswx_hls.yaml
Comment (nemo794) [inline src/proteus/schemas/dswx_hls.yaml:82]: If this is a correct statements, please add a note here (and everywhere else that it is relevant) that class 201 (evergreen) will always be masked out. Thanks!

File: src/proteus/dswx_hls.py
Comment (nemo794) [inline src/proteus/dswx_hls.py:4295]: Is % cloud coverage out of the entire 3660x3660 image? Or % coverage of the valid pixels?

This is perhaps a larger conversation with the group, but I'm starting to have reservations about these new metrics.

For one of the San Francisco Bay granules, here are the numbers:

```
  INPUT_HLS_PRODUCT_CLOUD_COVERAGE=25
  INPUT_HLS_PRODUCT_SPATIAL_COVERAGE=99
  CLOUD_COVERAGE=17
  SPATIAL_COVERAGE=52
```
I'm still trying to put my finger on why this is unsettling, but putting on my user hat, a couple of things stand out right off the bat:
1) Where did the clouds go?
2) Why did we lose half of the granule? To me, this has been a fantastic granule to use as a test case. But when I run queries to STAC-CMR, as a starting point I filter out anything with spatial coverage under 70%, which means I never would have found this granule.

Here's the browse image, for reference:
OPERA_DSWx-HLS_L8_30_T10SEG_20220119T184616Z_20230130T203109Z_v1 0 0_BROWSE

What are your thoughts on this? Is it still the right thing to do?

**Review by nemo794 — CHANGES_REQUESTED** [review]
Hi @gshiroma , Thank you for updating the version number and adding the new metadata.

As you'll see in the comments, I'm a bit concerned about a couple of the new metrics, and was quite confused by splitting the computation of the cloud layer into two parts. The first might require a larger conversation with the group, and the second I provided a suggestion for how to streamline the code a bit better.

After an initial look at the outputs/metadata, everything seems ok, but I'll take a deeper look after these changes are implemented in case they introduce new bugs. Thanks!

File: src/proteus/dswx_hls.py
Comment (gshiroma) [reply src/proteus/dswx_hls.py:1710]: Thank you, @nemo794 . The output is really `preliminary_cloud_layer`, meaning preliminary CLOUD layer (not WTR). I'll update the docstrings to make it more clear.

File: src/proteus/dswx_hls.py
Comment (gshiroma) [reply src/proteus/dswx_hls.py:1730]: I'll update the docstrings. Thank you, @nemo794

File: src/proteus/dswx_hls.py
Comment (gshiroma) [reply src/proteus/dswx_hls.py:1706]: The string value "mask" comes from the runconfig schema. I'd prefer to not change it now.

File: src/proteus/dswx_hls.py
Comment (gshiroma) [reply src/proteus/dswx_hls.py:1753]: Thank you for you suggestion. I'm trying to change the code as minimum as possible. I had to divide the original function into two parts because cloud/cloud shadow values were needed at the beginning of the execution (to populate the CLOUD_COVERAGE metadata of all DSWx-HLS layer) and the second part that needed the WTR-2 layer. I'd like to avoid adding another division to the original function.  To avoid confusion, I'm renaming `_compute_and_apply_cloud_layer()` to `_add_snow_to_cloud_layer_and_apply_cloud_masking()`.

File: src/proteus/dswx_hls.py
Comment (gshiroma) [reply src/proteus/dswx_hls.py:1767]: I'm updating the docstrings. "cover" comes from the runconfig schema and I'd prefer to avoid changing it now.

File: src/proteus/dswx_hls.py
Comment (gshiroma) [reply src/proteus/dswx_hls.py:4235]: I'm keeping the unpacking of the `image_dict()` at one place. I'm also adding a line to remove that dictionary when it's no longer needed: `del image_dict` after the line you pointed out above.

File: src/proteus/dswx_hls.py
Comment (gshiroma) [reply src/proteus/dswx_hls.py:4274]: Thank you, @nemo794 . The` mask_adjacent_to_cloud_mode`  `cover` requires the WTR-2 layer to update the CLOUD layer with snow/ice values. I update the docstrings and answered some of your comments above. If it's still unclear, we can meet tomorrow (Tuesday Jan 31st) to solve this issue.

File: src/proteus/dswx_hls.py
Comment (gshiroma) [reply src/proteus/dswx_hls.py:3525]: Uhm, `FOREST_MASK_LANDCOVER_CLASSES` has nothing to do with ocean masking.

File: src/proteus/dswx_hls.py
Comment (gshiroma) [reply src/proteus/dswx_hls.py]: Good point, @nemo794 . You are right, the HLS product defines the `CLOUD_COVERAGE` over valid pixels. I'm updating the code to do the same.

File: src/proteus/dswx_hls.py
Comment (gshiroma) [reply src/proteus/dswx_hls.py]: > Hi @gshiroma , Thank you for updating the version number and adding the new metadata.
> 
> As you'll see in the comments, I'm a bit concerned about a couple of the new metrics, and was quite confused by splitting the computation of the cloud layer into two parts. The first might require a larger conversation with the group, and the second I provided a suggestion for how to streamline the code a bit better.
> 
> After an initial look at the outputs/metadata, everything seems ok, but I'll take a deeper look after these changes are implemented in case they introduce new bugs. Thanks!

Thank you so much for your quick review, @nemo794 ! I answered/addressed all your comments. The only remaining comments are related to the evergreen class 201 that I forward the question to John. Please, let me know if you have any other comments or concerns.

File: src/proteus/dswx_hls.py
Comment (nemo794) [reply src/proteus/dswx_hls.py:3539]: Ah - my apologies - I highlighted the incorrect lines.

Currently, for the product metadata, the block of code on lines 3484-3485 will always set `OCEAN_MASKING_SHORELINE_DISTANCE_KM` to `ocean_masking_shoreline_distance_km`. But, what happens if there is no shoreline shape file provided? Is it correct to still set the metadata field `OCEAN_MASKING_SHORELINE_DISTANCE_KM` to the default number, e.g. `1` for 1 km?

Or, if there is no shoreline shapefile provided, would it be better to set `OCEAN_MASKING_SHORELINE_DISTANCE_KM` to `-`? This approach would be similar to how the code is handling `SHADOW_MASKING_ALGORITHM` when the alternate algorithm is used ([link](https://github.com/gshiroma/PROTEUS/blob/32d1a1315b4ecb2886267f013d1ccea1d956c412/src/proteus/dswx_hls.py#L3527-L3528)).

Comment (gshiroma) [reply src/proteus/dswx_hls.py]: Sure, sounds good, @nemo794 . I updated the code to assign `-` to the metadata field `OCEAN_MASKING_SHORELINE_DISTANCE_KM` if the shoreline shapefiile is not provided.

File: src/proteus/dswx_hls.py
Comment (nemo794) [inline src/proteus/dswx_hls.py:4391]: I think there's an issue with how the no data value is being set (and colored) for the LAND layer.

Running `gdalinfo` on the LAND.tif layer, the normal `NoData Value=...` field is missing, and the colortable entry assigned for value 255 is `255: 255,255,255,255`. (The other layers have this set as `255: 0,0,0,0`).

I'm not sure what's happening; could you please look into this? The Product Spec document says that there should be a fill value for this layer.

File: src/proteus/dswx_hls.py
Comment (nemo794) [inline src/proteus/dswx_hls.py:4332]: Git is not allowing me to comment on the relevant lines, but could you please confirm that it is correct that there is no fill data value for the SHAD.tif layer?

The Product Spec is not specific about this, and running `gdalinfo` on one of the *SHAD.tif output files shows that there is no fill data value specified, and then in the color table 255 is fully visible (not transparent). Thanks!

Comment (gshiroma) [reply src/proteus/dswx_hls.py:4332]: Hi @nemo794 , thank you for checking, but I think the shadow layer doesn't have fill value. We expect the DEM to be available globally. If it's not available for the area, the `_check_ancillary_inputs()` should raise an error.

Comment (nemo794) [reply src/proteus/dswx_hls.py:4365]: Sounds good! And thank you for updating the Product Spec document to match. 🚀

File: src/proteus/dswx_hls.py
Comment (nemo794) [reply src/proteus/dswx_hls.py:3572]: Per meeting with J. Jones, class `201` should not be included. (This was a misunderstanding on my part.)

File: src/proteus/dswx_hls.py
Comment (nemo794) [inline src/proteus/dswx_hls.py:3571]: If `forest_mask_landcover_classes` is set to `null` (or an empty list) in the runconfig, it looks like this metadata field would be empty. Should it be set to a dash `-` instead?

Comment (gshiroma) [reply src/proteus/dswx_hls.py]: Sounds good, @nemo794 . Just updated "-" rather than "" to empty `FOREST_MASK_LANDCOVER_CLASSES`. Thanks

File: src/proteus/dswx_hls.py
Comment (nemo794) [inline src/proteus/dswx_hls.py:1719]: Could the docstring please be updated to something like:

```suggestion
            Preliminary cloud mask (without the snow/ice class)
            with dtype of uint8 and values assigned as follows:
                0: Not masked
                1: Cloud shadow or adjacent to cloud/cloud shadow
                4: Cloud
                5: Cloud and class 1 (cloud shadow or adjacent to cloud/cloud shadow)
                255: Fill value (no data)

       See Also
       ---------
       _add_snow_to_cloud_layer : Add the snow bit encodings from `Fmask` to `preliminary_cloud_layer`
```
numpy docstring reference: [link](https://numpydoc.readthedocs.io/en/latest/format.html#see-also)

Note: the values I wrote above might be incorrect. But, this is why the docstring will be very helpful. Thanks!

File: src/proteus/dswx_hls.py
Comment (nemo794) [inline src/proteus/dswx_hls.py:1776]: Something like:

```suggestion
            Preliminary cloud mask (without the snow/ice class)
            with dtype of uint8 and values assigned as follows:
                0: Not masked
                1: Cloud shadow or adjacent to cloud/cloud shadow
                4: Cloud
                5: Cloud and class 1 (cloud shadow or adjacent to cloud/cloud shadow)
                255: Fill value (no data)
```

File: src/proteus/dswx_hls.py
Comment (gshiroma) [reply src/proteus/dswx_hls.py:1716]: Just update the docstrings. Thank you, @nemo794

File: src/proteus/dswx_hls.py
Comment (gshiroma) [reply src/proteus/dswx_hls.py:1782]: Just update the docstrings. Thank you, @nemo794

File: src/proteus/dswx_hls.py
Comment (nemo794) [inline src/proteus/dswx_hls.py:1795]:
```suggestion
              Cloud mask. This is the finished DSWx-HLS `CLOUD` layer,
              with dtype of uint8 and values assigned as follows:
                  0: Not masked
                  1: Cloud shadow or adjacent to cloud/cloud shadow
                  2: Snow/ice
                  3: Snow/ice and class 1 (cloud shadow or adjacent to cloud/cloud shadow)
                  4: Cloud
                  5: Cloud and class 1 (cloud shadow or adjacent to cloud/cloud shadow)
                  6: Cloud and snow/ice
                  7: Cloud, snow/ice, and class 1 (cloud shadow or adjacent to cloud/cloud shadow)
                  255: Fill value (no data)

See Also
---------
`_compute_preliminary_cloud_layer` : Can use to create input `cloud_layer`
```

File: src/proteus/dswx_hls.py
Comment (nemo794) [reply src/proteus/dswx_hls.py]: Thank you for updating `CLOUD COVERAGE`.

Regarding `SPATIAL_COVERAGE`, I'll just note here that as discussed in yesterday's meeting, the decision is to keep `SPATIAL_COVERAGE` computed as-is, and then add a new metadata field that computes spatial coverage over pixels that are not marked as "ocean" by the ocean mask.

File: src/proteus/dswx_hls.py
Comment (nemo794) [inline src/proteus/dswx_hls.py:1771]: Currently, this docstring says that the function takes in an argument named `cloud_layer`, and then returns an argument named `cloud_layer`. It's not great practice to use the same variable name for inputs and outputs. Plus, in the code itself, it looks like the input `cloud_layer` is being modified directly and so there's no need for it to be returned.

Two suggestions for how to approach this:
1) change the input variable name to be `preliminary_cloud_layer`, then make a copy of it in the code and return `cloud_layer`. This is the style of other functions in the codebase, and much clearer about what the input should be.
2) update the docstring to say that the input cloud layer is modified by the function, make sure to detail _how_ the function is modifying the actual values of the input layer, and also remove the `Return` statement. This strategy is less preferable to me, as it is contrary to the coding style used elsewhere in the code.

Comment (gshiroma) [reply src/proteus/dswx_hls.py:1771]: I prefer to avoid making an unnecessary copies of the array. I think the code "as is" is easier to follow.

File: src/proteus/dswx_hls.py
Comment (nemo794) [reply src/proteus/dswx_hls.py:1716]: Thank you for filling out the docstring. I'm not sure if something got mangled in the git merge, but now the output variable name and type disappeared, the indentation is off, and there should be an extra newline between the `Return` block and the `See Also` block. Could you please look into this? Thank you!

Comment (nemo794) [reply src/proteus/dswx_hls.py]: Your recent commit just fixed the spacing in the `_add_snow_to_cloud_layer` docstring (thank you!), but the spacing issues mentioned are still present in the `_compute_preliminary_cloud_layer` docstring. Could you please update? Thank you!

File: src/proteus/dswx_hls.py
Comment (nemo794) [reply src/proteus/dswx_hls.py:1773]: In that case, could you please follow suggestion 2?

Comment (gshiroma) [reply src/proteus/dswx_hls.py:1773]: I think the way it is now it's easier to understand. `cloud_layer` is mutable, but by returning it, we are making explicit that that's the output of the function. The input `preliminary_cloud_layer` and output `cloud_layer` are the same array, but I think the code this way is easier to follow. I updated the docstrings to make this point more clear:
```       cloud_layer : numpy.ndarray
            Preliminary cloud mask (without the snow/ice class). This array
            is updated within the function (mutable numpy.ndarray) and returned
            after the snow/ice class is added to the layer.```

Comment (nemo794) [reply src/proteus/dswx_hls.py:1773]: I agree that it's better to be explicit, both for the input arguments and the output arguments. In that case, please go with suggestion 1. There are so many other places in the code where extra copies of arrays are being created, I don't think that one additional occurance of that is going to matter.

Comment (gshiroma) [reply src/proteus/dswx_hls.py:1773]: I don't see a reason to add an unnecessary copy of data. The other array copies in the code are needed to create different layers. If that's not the case, please let me know because then we should fix that too.

Comment (nemo794) [reply src/proteus/dswx_hls.py:1773]: Is there a third approach? The requirements for the new approach would be that this function can not both mutate an input array and return that same array. It must only do one of these things. This is significantly cleaner.

Some pros and cons of each option: Mutating an input array makes for code that is very hard to debug and understand (as an example, mutating input arguments was a significant reason why the original QA code was nearly impossible to understand), but it avoids copying an array. Copying the array is much cleaner for the user to understand and is one of the foundations of functional programming, but does create a (temporary) copy of the array.

Comment (gshiroma) [reply src/proteus/dswx_hls.py:1773]: I guess the third approach is what we are doing lol. The behavior is described in the docstrings and also, this function intended for internal use (startswith `_`), so it won't be used by other modules. It's also very specific to a part of the code (add snow class to the cloud layer). Finally, having `preliminary_cloud_layer` and `cloud_layer` make it clear what is preliminary and final.

File: src/proteus/dswx_hls.py
Comment (nemo794) [inline src/proteus/dswx_hls.py:1184]: This can probably be simplified with `numpy.isin` ([link](https://numpy.org/doc/stable/reference/generated/numpy.isin.html#numpy.isin)).

Comment (gshiroma) [reply src/proteus/dswx_hls.py]: Thank you, @nemo794 ! I didn't know that function, it definitely simplifies the logic

**Review by nemo794 — COMMENTED** [review]
Once again, several great new features in this PR, including some complicated but vital improvements to the algorithm and code. I have several concerns about the coding style in terms of long term readability and maintainability, but the code seems to work.

There are many tweaks to the algorithm are still occurring offline, but I'll push these comments now. Thanks!

File: src/proteus/dswx_hls.py
Comment (gshiroma) [reply src/proteus/dswx_hls.py:1967]: This is the bit encoding. I'm updating the docstrings to make it more clear.

File: src/proteus/dswx_hls.py
Comment (gshiroma) [reply src/proteus/dswx_hls.py:3685]: Yeah, the problem is that, for example, the shoreline shapefile can be provided but not used, that's why I changed it from `not provided` to `-` (which BTW was already used for some other metadata fields). Any suggestions?

File: src/proteus/dswx_hls.py
Comment (gshiroma) [reply src/proteus/dswx_hls.py]: Sorry, I'm having a hard time figuring out what is wrong here. The indentation matches "Returns". Is it possible that this comment is from a previous version of the code?

File: src/proteus/dswx_hls.py
Comment (gshiroma) [reply src/proteus/dswx_hls.py]: I'm not finding these lines. Is it possible that this comment is from a previous version of the code?

File: src/proteus/dswx_hls.py
Comment (gshiroma) [reply src/proteus/dswx_hls.py:1948]: As discussed in the other thread, copying the array without necessity is out of discussion. I also think it's clearer to have `preliminary_cloud_layer` separated from the `cloud_layer`. The reader can see immediately what's preliminary and what's not. I can change it to have this PR merged, but I still prefer the current implementation compared to the two options that were presented.

File: src/proteus/dswx_hls.py
Comment (gshiroma) [reply src/proteus/dswx_hls.py]: `preliminary_cloud_layer` and `cloud_layer` are the same array, so the former doesn't take any extra memory.

**Comment by gshiroma** [comment]
> Once again, several great new features in this PR, including some complicated but vital improvements to the algorithm and code. I have several concerns about the coding style in terms of long term readability and maintainability, but the code seems to work.
> 
> There are many tweaks to the algorithm are still occurring offline, but I'll push these comments now. Thanks!

Thank you, @nemo794, for your review. I addressed your comments above. Please, let me know if you still have any other comments or concerns.

**Review by nemo794 — APPROVED** [review]
Code works! Will leave it to you to decide how to handle any of the remaining comments. Thanks again!

**Comment by gshiroma** [comment]
> 

Thank you, @nemo794 !

**Review by hfattahi — APPROVED** [review]
Thanks @gshiroma. Based on the sample products produced with this PR, I think we can move forward and merge these changes.

**Comment by gshiroma** [comment]
Thank you, @hfattahi

**Pull request merged**
